worker.go

```diff
@@ -76,48 +109,45 @@ func (w *Worker) Work() {
 func (w *Worker) ProcessMetric(m *UDPMetric) {
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
-	w.metrics++
+
+	w.processed++
+	w.wm.Upsert(m.MetricKey, m.Tags)
+
 	switch m.Type {
 	case "counter":
-		_, present := w.wm.counters[m.MetricKey]
-		if !present {
-			w.logger.WithField("name", m.Name).Debug("New counter")
-			w.wm.counters[m.MetricKey] = NewCounter(m.Name, m.Tags)
-		}
 		w.wm.counters[m.MetricKey].Sample(m.Value.(float64), m.SampleRate)
 	case "gauge":
-		_, present := w.wm.gauges[m.MetricKey]
-		if !present {
-			w.logger.WithField("name", m.Name).Debug("New gauge")
-			w.wm.gauges[m.MetricKey] = NewGauge(m.Name, m.Tags)
-		}
 		w.wm.gauges[m.MetricKey].Sample(m.Value.(float64), m.SampleRate)
 	case "histogram":
-		_, present := w.wm.histograms[m.MetricKey]
-		if !present {
-			w.logger.WithField("name", m.Name).Debug("New histogram")
-			w.wm.histograms[m.MetricKey] = NewHist(m.Name, m.Tags)
-		}
 		w.wm.histograms[m.MetricKey].Sample(m.Value.(float64), m.SampleRate)
 	case "set":
-		_, present := w.wm.sets[m.MetricKey]
-		if !present {
-			w.logger.WithField("name", m.Name).Debug("New set")
-			w.wm.sets[m.MetricKey] = NewSet(m.Name, m.Tags)
-		}
 		w.wm.sets[m.MetricKey].Sample(m.Value.(string), m.SampleRate)
 	case "timer":
-		_, present := w.wm.timers[m.MetricKey]
-		if !present {
-			w.logger.WithField("name", m.Name).Debug("New timer")
-			w.wm.timers[m.MetricKey] = NewHist(m.Name, m.Tags)
-		}
 		w.wm.timers[m.MetricKey].Sample(m.Value.(float64), m.SampleRate)
 	default:
 		w.logger.WithField("type", m.Type).Error("Unknown metric type")
 	}
 }
 
+func (w *Worker) ImportMetric(other JSONMetric) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	// we don't increment the processed metric counter here, it was already
+	// counted by the original veneur that sent this to us
+	w.imported++
+	w.wm.Upsert(other.MetricKey, other.Tags)
+
+	switch other.Type {
+	case "set":
+		if err := w.wm.sets[other.MetricKey].Combine(other.Value); err != nil {
+			w.logger.WithError(err).Error("Could not merge sets")
+		}
+	default:
+		w.logger.WithField("type", other.Type).Error("Unknown metric type")
```

Comment:
For whomever has to see this error, perhaps improve it to say what it's trying to do (import/merge/whatever?) when it encountered this?